test/test_tagging.py

```diff
@@ -0,0 +1,232 @@
+import json
+from typing import (
+    Mapping,
+    Optional,
+    Tuple,
+)
+import unittest
+from unittest.mock import (
+    MagicMock,
+    patch,
+)
+
+from azul.deployment import (
+    populate_tags,
+)
+from azul.json import (
+    AnyJSON,
+)
+from azul.logging import (
+    configure_test_logging,
+)
+from azul_test_case import (
+    AzulUnitTestCase,
+)
+
+
+def setupModule():
+    configure_test_logging()
+
+
+class TestTerraformResourceTags(AzulUnitTestCase):
+
+    def assertDictEqualPermissive(self, expected: Mapping, actual: Mapping):
+        path = self.permissive_compare(expected, actual)
+        self.assertIsNone(path, f'Discrepancy in path: {path}')
+
+    def permissive_compare(self,
+                           expected: AnyJSON,
+                           actual: AnyJSON,
+                           *path: str) -> Optional[Tuple[str]]:
+        """
+        Recursive dictionary comparison. Skips comparison for value in
+        `actual` when the corresponding value in `expected` is set to
+        None.
+
+        Returns None if dictionaries match, or path of discrepancy as a
+        tuple otherwise.
+
+        >>> t = TestTerraformResourceTags()
+        >>> t.permissive_compare(
+        ...     {'foo': 'bar', 'qaz': [{'qux': 123, 456: 789}]},
+        ...     {'foo': 'bar', 'qaz': [{'qux': 123, 456: 789}]}
+        ... )
+
+        >>> t.permissive_compare(
+        ...     {'foo': 'bar', 'qaz': [{'qux': 123, 456: None}]},
+        ...     {'foo': 'bar', 'qaz': [{'qux': 123, 456: 'abc'}]}
+        ... )
+
+        >>> t.permissive_compare(
+        ...     {'foo': 'bar', 'qaz': [{'qux': 123, 456: 'def'}]},
+        ...     {'foo': 'bar', 'qaz': [{'qux': 123, 456: 'abc'}]}
+        ... )
+        ...
+        ('qaz', '[0]', '456')
+
+        >>> t.permissive_compare(
+        ...     {'foo': 'bar', 'qaz': [{'qux': 123, 456: 'def'}]},
+        ...     {'foo': 'bar', 'qaz': [{'qux': 123, 456: None}]}
+        ... )
+        ('qaz', '[0]', '456')
+
+        >>> t.permissive_compare(
+        ...     {'foo': 'bar', 'qaz': None, 456: 'def'},
+        ...     {'foo': 'bar', 'qaz': [{'qux': 123, 456: 'abc'}]}
+        ... )
+        ('456',)
+
+        >>> t.permissive_compare(
+        ...     {'foo': 'bar', 'qaz': [{'qux': {123: 890}}]},
+        ...     {'foo': 'bar', 'qaz': [{'qux': {123: 456}}]}
+        ... )
+        ('qaz', '[0]', 'qux', '123')
+        """
+        if isinstance(actual, list) and isinstance(expected, list):
+            for i, v in enumerate(zip(expected, actual)):
+                diff = self.permissive_compare(*v, *path, f'[{i}]')
```

Comment:
We can avoid inventing syntax by relaxing the type constraint on keys to `Union[str, int]`